src/forwarder.rs

```diff
@@ -0,0 +1,137 @@
+/*
+ * This file is part of Edgehog.
+ *
+ * Copyright 2023 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Manage the device forwarder operation.
+
+use std::borrow::Borrow;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    hash::Hash,
+    ops::Deref,
+};
+
+use astarte_device_sdk::{Aggregation, AstarteDeviceDataEvent};
+use edgehog_forwarder::astarte::{retrieve_connection_info, ConnectionInfo};
+use edgehog_forwarder::connections_manager::ConnectionsManager;
+use log::error;
+use reqwest::Url;
+use tokio::task::JoinHandle;
+
+#[derive(Debug)]
+struct Key(ConnectionInfo);
```

Comment:
In this case you cannot use ConnectionInfo directly?  and Add the `impl traits` directly to ConnectionInfo? you often use connectionInfo so I think `Key` struct it's extra.